File: js/src/gdb/tests/test-jsval.py
# Basic unit tests for jsval pretty-printer.
# flake8: noqa: F821

assert_subprinter_registered("SpiderMonkey", "JS::Value")

run_fragment("jsval.simple")

assert_pretty("fortytwo", "$JS::Int32Value(42)")
assert_pretty("fortytwoD", "$JS::DoubleValue(42.0)")
assert_pretty("negone", "$JS::Int32Value(-1)")
assert_pretty("undefined", "$JS::UndefinedValue()")
assert_pretty("null", "$JS::NullValue()")
assert_pretty("js_true", "$JS::BooleanValue(true)")
assert_pretty("js_false", "$JS::BooleanValue(false)")
assert_pretty("elements_hole", "$JS::MagicValue(JS_ELEMENTS_HOLE)")
assert_pretty("empty_string", '$JS::Value("")')
assert_pretty("friendly_string", '$JS::Value("Hello!")')
assert_pretty("symbol", '$JS::Value(Symbol.for("Hello!"))')
assert_pretty("bi", "$JS::BigIntValue()")
assert_pretty("global", "$JS::Value((JSObject *)  [object global])")
assert_pretty(
    "onehundredthirtysevenonehundredtwentyeighths", "$JS::DoubleValue(1.0703125)"
)

File: js/src/gdb/tests/test-jsval.cpp
#include "gdb-tests.h"

#include "js/GlobalObject.h"
#include "js/String.h"
#include "js/Symbol.h"
#include "vm/BigIntType.h"

FRAGMENT(jsval, simple) {
  using namespace JS;

  RootedValue fortytwo(cx, Int32Value(42));
  RootedValue fortytwoD(cx, DoubleValue(42));
  RootedValue negone(cx, Int32Value(-1));
  RootedValue undefined(cx, UndefinedValue());
  RootedValue null(cx, NullValue());
  RootedValue js_true(cx, BooleanValue(true));
  RootedValue js_false(cx, BooleanValue(false));
  RootedValue elements_hole(cx, js::MagicValue(JS_ELEMENTS_HOLE));

  RootedValue empty_string(cx);
  empty_string.setString(JS_NewStringCopyZ(cx, ""));
  RootedString hello(cx, JS_NewStringCopyZ(cx, "Hello!"));
  RootedValue friendly_string(cx, StringValue(hello));
  RootedValue symbol(cx, SymbolValue(GetSymbolFor(cx, hello)));
  RootedValue bi(cx, BigIntValue(BigInt::zero(cx)));

  RootedValue global(cx);
  global.setObject(*CurrentGlobalOrNull(cx));

  // Some interesting value that floating-point won't munge.
  RootedValue onehundredthirtysevenonehundredtwentyeighths(
      cx, DoubleValue(137.0 / 128.0));

  breakpoint();

  use(fortytwo);
  use(fortytwoD);
  use(negone);
  use(undefined);
  use(js_true);
  use(js_false);
  use(null);
  use(elements_hole);
  use(empty_string);
  use(friendly_string);
  use(symbol);
  use(bi);
  use(global);
}
